### PR TITLE
feat(ui-studies): add on click fetch and display list of non studies folder

### PR DESCRIPTION
**Pull request by smailio**

### CODE REVIEW

File: webapp/src/components/App/Studies/utils.ts [hunk 1]
```diff
@@ -20,6 +20,13 @@ export interface StudyTreeNode {
   children: StudyTreeNode[];
 }
 
+export interface NonStudyFolder {
```
Comment (skamril) [inline webapp/src/components/App/Studies/utils.ts]: It is better to follow the types from the API for maintainability:

```typescript
// We add 'DTO' prefix normally but the backend didn't add it
export interface NonStudyFolder {
  name: string;
  path: string;
  workspace: string;
}

// New type for the frontend
export interface NonStudyFolderWithParentPath extends NonStudyFolder {
  parentPath: string;
}
```

File: webapp/src/services/api/study.ts [hunk 2]
```diff
@@ -48,6 +49,29 @@ export const getStudies = async (): Promise<StudyMetadata[]> => {
   });
 };
 
+export const getFolders = async (
+  workspace: string,
+  folderName: string,
+): Promise<NonStudyFolder[]> => {
+  const res = await client.get(
```
Comment (skamril) [inline webapp/src/services/api/study.ts:56]:
```typescript
await client.get<NonStudyFolder>( // Specify return type
  `/v1/private/explorer/${workspace}/_list_dir`, 
  { params: { folderName } }, // Enconding by Axios automatically
);
```
Then `map((folder: NonStudyFolder) =>` -> `map((folder) =>`

File: webapp/src/services/api/study.ts [hunk 3]
```diff
@@ -48,6 +49,29 @@ export const getStudies = async (): Promise<StudyMetadata[]> => {
   });
 };
 
+export const getFolders = async (
+  workspace: string,
+  folderName: string,
+): Promise<NonStudyFolder[]> => {
```
Comment (skamril) [inline webapp/src/services/api/study.ts]: Following my previous comment:

`Promise<NonStudyFolderWithParentPath[]>`

File: webapp/src/services/api/study.ts [hunk 4]
```diff
@@ -48,6 +49,29 @@ export const getStudies = async (): Promise<StudyMetadata[]> => {
...
+  const res = await client.get(
+    `/v1/private/explorer/${workspace}/_list_dir?path=${encodeURIComponent(folderName)}`,
+  );
+  return res.data.map((folder: NonStudyFolder) => {
+    const parentPath = [
+      "",
+      folder.workspace,
+      ...folder.path
+        .split("/")
+        .filter((elm: string) => elm !== "")
```
Comment (skamril) [inline webapp/src/services/api/study.ts]: Tip: `.filter(Boolean)`

File: webapp/src/components/App/Studies/StudyTree.tsx [hunk 5]
```diff
@@ -21,18 +21,127 @@ import TreeItemEnhanced from "../../common/TreeItemEnhanced";
 import { SimpleTreeView } from "@mui/x-tree-view/SimpleTreeView";
 import { getParentPaths } from "../../../utils/pathUtils";
 import * as R from "ramda";
+import { useEffect, useState } from "react";
+import * as api from "../../../services/api/study";
+
+/**
+ * Add a folder that was returned by the explorer into the study tree view.
+ *
+ * This folder isn't a study, otherwise it woudl'nt be
```
Comment (skamril) [inline webapp/src/components/App/Studies/StudyTree.tsx]: "wouldn't"

File: webapp/src/components/App/Studies/StudyTree.tsx [hunk 6]
```diff
@@ -21,18 +21,127 @@ import TreeItemEnhanced from "../../common/TreeItemEnhanced";
...
+  const emptyNode: StudyTreeNode = { name: "", path: "", children: [] };
+  const [studiesTree, setStudiesTree] = useState(emptyNode);
+  const [initializedFlag, setInitializedFlag] = useState(false);
   const folder = useAppSelector((state) => getStudyFilters(state).folder, R.T);
-  const studiesTree = useAppSelector(getStudiesTree);
+  const initialStudiesTree = useAppSelector(getStudiesTree);
   const dispatch = useAppDispatch();
 
+  useEffect(() => {
+    if (!initializedFlag) {
```
Comment (skamril) [inline webapp/src/components/App/Studies/StudyTree.tsx]: Why you don't update when the `studiesTree` change after the first initialization?

Otherwise you can do this and remove the `useEffect`:

```typescript
const initialStudiesTree = useAppSelector(getStudiesTree);
const [studiesTree, setStudiesTree] = useState(initialStudiesTree);
```

File: webapp/src/components/App/Studies/StudyTree.tsx [hunk 7]
```diff
@@ -21,18 +21,127 @@ import TreeItemEnhanced from "../../common/TreeItemEnhanced";
...
+    }
+  }, [studiesTree]);
+
   ////////////////////////////////////////////////////////////////
   // Event Handlers
   ////////////////////////////////////////////////////////////////
 
   const handleTreeItemClick = (itemId: string) => {
     dispatch(updateStudyFilters({ folder: itemId }));
+    fetchAndMergeSubfolders(itemId, studiesTree).then((nextStudiesTree) =>
```
Comment (skamril) [inline webapp/src/components/App/Studies/StudyTree.tsx]: Tip: `.then(setStudiesTree)`

**Review by skamril — CHANGES_REQUESTED** [review]

File: webapp/src/components/App/Studies/StudyTree.test.tsx
Comment (hdinia) [inline webapp/src/components/App/Studies/StudyTree.test.tsx:1]: since you added this test file, please consider making a folder named like the component and place files inside:

/Studies/StudyTree/index.tsx
/Studies/StudyTree/index.test.tsx

Comment (hdinia) [inline webapp/src/components/App/Studies/StudyTree.test.tsx:1]: The current test structure could benefit from using fixtures and parameterized tests. I suggest refactoring to:
- Extract test cases to a separate fixtures file
- Use `test.each` for similar test scenarios
- Add edge cases (empty trees, invalid paths)

This will make the tests more maintainable, easier to extend. See example implementation above.
```typescript
// fixtures.ts
export const FIXTURES = {
  basicTree: {
    name: "Basic tree with single level",
    studyTree: {
      name: "Root",
      path: "/",
      children: [
        { name: "a", path: "/a", children: [] },
        { name: "b", path: "/b", children: [] },
      ],
    },
    folders: [
      {
        name: "folder1",
        path: "folder1",
        workspace: "a",
        parentPath: "/a",
      },
    ],
    expected: {
      name: "Root",
      path: "/",
      children: [
        {
          name: "a",
          path: "/a",
          children: [{ name: "folder1", path: "/a/folder1", children: [] }],
        },
        { name: "b", path: "/b", children: [] },
      ],
    },
  },
  nestedTree: {
    name: "Nested tree structure",
    studyTree: {
      name: "Root",
      path: "/",
      children: [
        {
          name: "a",
          path: "/a",
          children: [{ name: "suba", path: "/a/suba", children: [] }],
        },
      ],
    },
    folders: [
      {
        name: "folder1",
        path: "suba/folder1",
        workspace: "a",
        parentPath: "/a/suba",
      },
    ],
    expected: {
      name: "Root",
      path: "/",
      children: [
        {
          name: "a",
          path: "/a",
          children: [
            {
              name: "suba",
              path: "/a/suba",
              children: [
                { name: "folder1", path: "/a/suba/folder1", children: [] },
              ],
            },
          ],
        },
      ],
    },
  },
  duplicateCase: {
    name: "Tree with potential duplicates",
    studyTree: {
      name: "Root",
      path: "/",
      children: [
        {
          name: "a",
          path: "/a",
          children: [{ name: "folder1", path: "/a/folder1", children: [] }],
        },
      ],
    },
    folders: [
      {
        name: "folder1",
        path: "/folder1",
        workspace: "a",
        parentPath: "/a",
      },
    ],
    expected: {
      name: "Root",
      path: "/",
      children: [
        {
          name: "a",
          path: "/a",
          children: [{ name: "folder1", path: "/a/folder1", children: [] }],
        },
      ],
    },
  },
  multipleFolders: {
    name: "Multiple folders merge",
    studyTree: {
      name: "Root",
      path: "/",
      children: [{ name: "a", path: "/a", children: [] }],
    },
    folders: [
      {
        name: "folder1",
        path: "/folder1",
        workspace: "a",
        parentPath: "/a",
      },
      {
        name: "folder2",
        path: "/folder2",
        workspace: "a",
        parentPath: "/a",
      },
    ],
    expected: {
      name: "Root",
      path: "/",
      children: [
        {
          name: "a",
          path: "/a",
          children: [
            { name: "folder1", path: "/a/folder1", children: [] },
            { name: "folder2", path: "/a/folder2", children: [] },
          ],
        },
      ],
    },
  },
};
```
```typescript
// index.test.tsx
import { FIXTURES } from "./fixtures";

describe("StudyTree Utils", () => {
  describe("mergeStudyTreeAndFolders", () => {
    test.each(Object.values(FIXTURES))(
      "$name",
      ({ studyTree, folders, expected }) => {
        const result = mergeStudyTreeAndFolders(studyTree, folders);
        expect(result).toEqual(expected);
      },
    );

    test("should handle empty study tree", () => {
      const emptyTree: StudyTreeNode = {
        name: "Root",
        path: "/",
        children: [],
      };
      const result = mergeStudyTreeAndFolders(emptyTree, []);
      expect(result).toEqual(emptyTree);
    });

    test("should handle empty folders array", () => {
      const tree: StudyTreeNode = {
        name: "Root",
        path: "/",
        children: [{ name: "a", path: "/a", children: [] }],
      };
      const result = mergeStudyTreeAndFolders(tree, []);
      expect(result).toEqual(tree);
    });

    test("should handle invalid parent paths", () => {
      const tree: StudyTreeNode = {
        name: "Root",
        path: "/",
        children: [{ name: "a", path: "/a", children: [] }],
      };
      const invalidFolder: NonStudyFolder = {
        name: "invalid",
        path: "/invalid",
        workspace: "nonexistent",
        parentPath: "/nonexistent",
      };
      const result = mergeStudyTreeAndFolders(tree, [invalidFolder]);
      expect(result).toEqual(tree);
    });
  });
});
```

File: webapp/src/components/App/Studies/StudyTree.tsx [hunk 8]
```diff
@@ -21,18 +21,127 @@ import TreeItemEnhanced from "../../common/TreeItemEnhanced";
...
+ * returned by the explorer API, but this folder can have study in it
+ * and still not be in the initial study tree that's parsed from the study
+ * list, this happen when the studies in the folder aren't scanned yet.
+ *
+ * However we want to allow the user to see these folder. When the user explore
+ * they shoudln't wait for a long running scan to complete before they're able to
+ * a folder in the hierarchy.
+ *
+ * @param studiesTree
+ * @param folder
```
Comment (hdinia) [inline webapp/src/components/App/Studies/StudyTree.tsx]: look like the params descriptions are missing

File: webapp/src/components/App/Studies/StudyTree.tsx [hunk 9]
```diff
@@ -21,18 +21,127 @@ import TreeItemEnhanced from "../../common/TreeItemEnhanced";
...
+      ...studiesTree,
+      children: studiesTree.children.map((child) =>
+        mergeStudyTreeAndFolder(child, folder),
+      ),
+    };
+  } else {
+    // folder isn't part of this hierarchy
+    return studiesTree;
+  }
+}
```
Comment (hdinia) [inline webapp/src/components/App/Studies/StudyTree.tsx]: suggestion: prefer using early return synthax, and native methods over iterators 
Early returns:
- Reduces cognitive load by handling edge cases first
- Avoids deep nesting of if/else blocks
- Better performance by avoiding unnecessary code execution

Using native methods (some() vs for loop):
- More declarative and easier to understand intent
- Likely optimized by the JS engine vs manual iteration
- Reduces mutation risks and side effects
- Less boilerplate and temporary variables

```typescript
/**
 * Merges a non-study folder into the study tree hierarchy.
 * 
 * These folders may contain unscanned studies and should be visible
 * in the tree view without waiting for scan completion.
 * 
 * @param studyTree - The current study tree structure
 * @param folder - Non-study folder to merge
 * @returns Updated study tree with merged folder
 */
function mergeStudyTreeAndFolder(
  studyTree: StudyTreeNode,
  folder: NonStudyFolder,
): StudyTreeNode {
  // Early return if folder doesn't belong in this branch
  if (!folder.parentPath.startsWith(studyTree.path)) {
    return studyTree;
  }

  // Direct child case
  if (folder.parentPath === studyTree.path) {
    const folderExists = studyTree.children.some(child => child.name === folder.name);
    if (folderExists) return studyTree;

    return {
      ...studyTree,
      children: [
        ...studyTree.children,
        {
          path: `${folder.parentPath}/${folder.name}`,
          name: folder.name,
          children: [],
        },
      ],
    };
  }

  // Recursive case - update children
  return {
    ...studyTree,
    children: studyTree.children.map(child => 
      mergeStudyTreeAndFolder(child, folder)
    ),
  };
}
  ```

File: webapp/src/components/App/Studies/StudyTree.tsx [hunk 10]
```diff
@@ -21,18 +21,127 @@ import TreeItemEnhanced from "../../common/TreeItemEnhanced";
...
+export function mergeStudyTreeAndFolders(
+  studiesTree: StudyTreeNode,
+  folders: NonStudyFolder[],
+): StudyTreeNode {
+  let mergedStudyTree = studiesTree;
+  for (const folder of folders) {
+    mergedStudyTree = mergeStudyTreeAndFolder(mergedStudyTree, folder);
+  }
+  return mergedStudyTree;
+}
```
Comment (hdinia) [inline webapp/src/components/App/Studies/StudyTree.tsx]: suggestion: be careful about state mutations, here you can for example use reduce() method as it handles the fold operation more elegantly while maintaining immutability principles.

```typescript
export const mergeStudyTreeAndFolders = (
  studyTree: StudyTreeNode,
  folders: NonStudyFolder[],
): StudyTreeNode => folders.reduce(
  (tree, folder) => mergeStudyTreeAndFolder(tree, folder), 
  studyTree
);
```

File: webapp/src/components/App/Studies/StudyTree.tsx (hunk 6 above)
Comment (hdinia) [reply webapp/src/components/App/Studies/StudyTree.tsx]: @smailio the bug of the subfolders not appearing on the initial render comes from that, here's what I suggest to complete with Samir's comment:

```typescript
  const initialStudiesTree = useAppSelector(getStudiesTree);
  const [studiesTree, setStudiesTree] = useState(initialStudiesTree);
  const folder = useAppSelector((state) => getStudyFilters(state).folder, R.T);
  const dispatch = useAppDispatch();

  // Initialize folders once we have the tree
  useEffect(() => {
    if (!folder || !initialStudiesTree.children.length) {
      return;
    }

    // extract folder paths excluding root
    const getFolderPaths = (path: string): string[] =>
      path.split("/").reduce<string[]>((acc, _, i, parts) => {
        const currentPath = parts.slice(0, i + 1).join("/");
        return currentPath !== "root" ? [...acc, currentPath] : acc;
      }, []);

    // sequential folder fetching with error handling
    const updateFolderTree = async () => {
      try {
        const finalTree = await getFolderPaths(folder).reduce(
          async (treePromise, path) => {
            const tree = await treePromise;
            return fetchAndMergeSubfolders(path, tree);
          },
          Promise.resolve(initialStudiesTree),
        );

        setStudiesTree(finalTree);
      } catch (error) {
        // here you can use a snackback error component if needed
        console.error("Failed to initialize folders:", error);
      }
    };

    updateFolderTree();
  }, [folder, initialStudiesTree]);
  ```

**Review by hdinia — CHANGES_REQUESTED** [review]

File: webapp/src/services/api/study.ts [hunk 11]
```diff
@@ -48,6 +49,29 @@ export const getStudies = async (): Promise<StudyMetadata[]> => {
   });
 };
 
+export const getFolders = async (
+  workspace: string,
+  folderName: string,
```
Comment (skamril) [inline webapp/src/services/api/study.ts]: It's only a name or can be a path?

File: webapp/src/components/App/Studies/StudyTree/index.tsx [hunk 12]
```diff
@@ -0,0 +1,298 @@
...
+ * However we want to allow the user to see these folder. When the user explore
+ * they shoudln't wait for a long running scan to complete before they're able to
+ * a folder in the hierarchy.
+ *
+ * @param studiesTree study tree to insert the folder into
+ * @param folder folder to inert into the tree
+ * @returns study tree with the folder inserted if it wasn't already there.
+ * New branch is created if it contain the folder otherwise the branch is left unchanged.
+ */
+function insertFolderIfNotExist(
```
Comment (skamril) [inline webapp/src/components/App/Studies/StudyTree/index.tsx]: Create a utils.ts for the functions in StudyTree folder

File: webapp/src/components/App/Studies/StudyTree/index.tsx [hunk 13]
```diff
@@ -0,0 +1,298 @@
...
+            );
+          }
+        });
+      })
+      .catch((error) => {
+        enqueueErrorSnackbar("Failed to load list workspaces", error);
+        setStudiesTree(initialStudiesTree);
+      });
+  }, [initialStudiesTree]);
+  ////////////////////////////////////////////////////////////////
```
Comment (skamril) [inline webapp/src/components/App/Studies/StudyTree/index.tsx:235]: Line break missing

File: webapp/src/components/App/Studies/StudyTree/index.tsx [hunk 14]
```diff
@@ -0,0 +1,298 @@
...
+  const handleTreeItemClick = (
+    itemId: string,
+    studyTreeNode: StudyTreeNode,
+  ) => {
+    dispatch(updateStudyFilters({ folder: itemId }));
+    if (itemId === "root") {
+      fetchAndInsertWorkspaces(studiesTree)
+        .then(setStudiesTree)
+        .catch((error) => {
+          enqueueErrorSnackbar("Failed to load list workspaces", error);
```
Comment (skamril) [inline webapp/src/components/App/Studies/StudyTree/index.tsx]: Add i18n key

File: webapp/src/components/App/Studies/StudyTree/index.tsx [hunk 15]
```diff
@@ -0,0 +1,298 @@
...
+    }
+    const chidrenPaths = studyTreeNode.children.map(
+      (child) => `root${child.path}`,
+    );
+    // children paths and current element path
+    fetchAndInsertSubfolders(chidrenPaths, studiesTree).then((r) => {
+      setStudiesTree(r[0]);
+      for (const path of r[1]) {
+        enqueueErrorSnackbar(
+          `Failed to initialize folders for : ${path}`,
```
Comment (skamril) [inline webapp/src/components/App/Studies/StudyTree/index.tsx]: You are not missing the `catch`? 

And add i18n key

File: webapp/src/components/App/Studies/utils.ts [hunk 16]
```diff
@@ -20,6 +20,16 @@ export interface StudyTreeNode {
   children: StudyTreeNode[];
 }
 
+export interface NonStudyFolderDTO {
```
Comment (skamril) [inline webapp/src/components/App/Studies/utils.ts:23]: Rename `class NonStudyFolder(AntaresBaseModel)` with 'DTO' suffix in Python code

**Review by skamril — CHANGES_REQUESTED** [review]

**Review by hdinia — CHANGES_REQUESTED** [review]
You have an import error in components/Studies/SideNav.tsx

update this `import StudyTree from "./StudyTree";`
to: `import StudyTree from "@/components/App/Studies/StudyTree";`


do you have all frontend tools properly setup, eslint, typescript etc ?

File: webapp/src/components/App/Studies/StudyTree/index.tsx [hunk 17]
```diff
@@ -0,0 +1,298 @@
...
+        const subfolders = await fetchSubfolders(path);
+        return [insertFoldersIfNotExist(accTree, subfolders), accFailedPaths];
+      } catch (error) {
+        console.error("failed to load path ", path, error);
+        return [accTree, [...accFailedPaths, path]];
+      }
+    },
+    Promise.resolve<[StudyTreeNode, string[]]>([studiesTree, []]),
+  );
+}
```
Comment (hdinia) [inline webapp/src/components/App/Studies/StudyTree/index.tsx]: The current code processes paths sequentially through an async reduce, forcing each path to wait for the previous one to complete. This creates unnecessary delays that compound with more paths. Error handling is also inefficient, with try/catch blocks cascading through the promise chain and complicating state management.

To avoid this you can use parallel fetching using `Promise.allSettled` instead of sequential async reduce

```typescript
async function fetchAndInsertSubfolders(
  paths: string[],
  studiesTree: StudyTreeNode,
): Promise<[StudyTreeNode, string[]]> {
  const results = await Promise.allSettled(
    paths.map((path) => fetchSubfolders(path)),
  );

  return results.reduce<[StudyTreeNode, string[]]>(
    ([tree, failed], result, index) => {
      if (result.status === "fulfilled") {
        return [insertFoldersIfNotExist(tree, result.value), failed];
      }
      console.error("Failed to load path:", paths[index], result.reason);
      return [tree, [...failed, paths[index]]];
    },
    [studiesTree, []],
  );
}
```

File: webapp/src/components/App/Studies/StudyTree/index.tsx [hunk 18]
```diff
@@ -0,0 +1,298 @@
...
+ * fetch and insert the workspaces into the study tree.
+ * @param studyTree study tree to insert the workspaces into
+ * @returns study tree with the workspaces inserted if they weren't already there.
+ */
+async function fetchAndInsertWorkspaces(
+  studyTree: StudyTreeNode,
+): Promise<StudyTreeNode> {
+  const workspaces = await api.getWorkspaces();
+  return insertWorkspacesIfNotExist(studyTree, workspaces);
+}
```
Comment (hdinia) [inline webapp/src/components/App/Studies/StudyTree/index.tsx]: you can remove intermediate functions, and streamline workspace insertion
```typescript
async function fetchAndInsertWorkspaces(
  studyTree: StudyTreeNode,
): Promise<StudyTreeNode> {
  const workspaces = await api.getWorkspaces();

  return workspaces.reduce((tree, workspace) => {
    const exists = tree.children.some((child) => child.name === workspace);
    
    if (exists) {
      return tree;
    }

    return {
      ...tree,
      children: [
        ...tree.children,
        { name: workspace, path: `/${workspace}`, children: [] },
      ],
    };
  }, studyTree);
}
```

File: webapp/src/components/App/Studies/StudyTree/index.tsx [hunk 19]
```diff
@@ -0,0 +1,298 @@
...
+              "details in console.error",
+            );
+          }
+        });
+      })
+      .catch((error) => {
+        enqueueErrorSnackbar("Failed to load list workspaces", error);
+        setStudiesTree(initialStudiesTree);
+      });
+  }, [initialStudiesTree]);
```
Comment (hdinia) [inline webapp/src/components/App/Studies/StudyTree/index.tsx]: Refactor folder initialization to use async/await instead of nested promise chains

please consider avoiding promise chaining, because it can lead to the "callback hell" very know issue of nested .then() handlers that make the flow hard to follow:

Also as stated in previous review please try to avoid iterators as opposed to native methods (unless you need that for a specific reason)
Native methods are optimized by JavaScript engines, they are declarative and widely understood syntax + they ensure immutability which aligns with React's principles

```typescript
  useUpdateEffectOnce(() => {
    const initializeFolders = async () => {
      try {
        const treeWithWorkspaces =
          await fetchAndInsertWorkspaces(initialStudiesTree);
        const childrenPaths = treeWithWorkspaces.children.map(
          (child) => `root${child.path}`,
        );

        const [updatedTree, failedPaths] = await fetchAndInsertSubfolders(
          childrenPaths,
          treeWithWorkspaces,
        );

        setStudiesTree(updatedTree);
        failedPaths.forEach((path) => {
          enqueueErrorSnackbar(
             // use dedicated translation key here
          );
        });
      } catch (error) {
        enqueueErrorSnackbar(
          // use dedicated translation key here
          error as AxiosError,
        );
        setStudiesTree(initialStudiesTree);
      }
    };

    initializeFolders();
  }, [initialStudiesTree]);
  ```

**Review by hdinia — CHANGES_REQUESTED** [review]

File: webapp/src/services/api/study.ts [hunk 20]
```diff
@@ -48,6 +49,37 @@ export const getStudies = async (): Promise<StudyMetadata[]> => {
   });
 };
 
+export const getWorkspaces = async (): Promise<string[]> => {
+  const res = await client.get(`/v1/private/explorer/_list_workspaces`);
+  return res.data.map((folder: { name: string }) => folder.name);
+};
```
Comment (hdinia) [inline webapp/src/services/api/study.ts]: you can pass types to axios methods, like this:

```typescript
interface Workspace {
  name: string;
}

export const getWorkspaces = async => {
  const res = await client.get<Workspace[]>(
    `/v1/private/explorer/_list_workspaces`,
  );
  return res.data.map((workspace) => workspace.name);
};
```

File: webapp/src/components/App/Studies/StudyTree/index.tsx (hunk 15 above)
Comment (smailio) [reply webapp/src/components/App/Studies/StudyTree/index.tsx]: No no the function return the [studyTree, failedPaths[]], the catch happen earlier in fetchAndInsertSubfolders

File: webapp/src/services/api/study.ts [hunk 21]
```diff
@@ -48,6 +49,37 @@ export const getStudies = async (): Promise<StudyMetadata[]> => {
...
+  return res.data.map((folder) => {
+    const parentPath = [
+      "",
+      folder.workspace,
+      ...folder.path.split("/").filter(Boolean).slice(0, -1),
+    ].join("/");
+    return {
+      ...folder,
+      parentPath,
+    };
```
Comment (hdinia) [inline webapp/src/services/api/study.ts]: Maybe this path manipulation logic should be on the API side ?

File: webapp/src/services/api/study.ts [hunk 22]
```diff
@@ -48,6 +49,37 @@ export const getStudies = async (): Promise<StudyMetadata[]> => {
   });
 };
 
+export const getWorkspaces = async (): Promise<string[]> => {
+  const res = await client.get(`/v1/private/explorer/_list_workspaces`);
```
Comment (hdinia) [inline webapp/src/services/api/study.ts]: Any specific reason why we have underscores in the api url ? `/v1/private/explorer/_list_workspaces`
We follow [REST conventions](https://restfulapi.net/resource-naming/), this should be something like `/v1/private/explorer/workspaces`, and if separators are needed we usually use a hyphen "-"

Comment (smailio) [reply webapp/src/services/api/study.ts]: It's as per the specs, convention for private routes

**Review by MartinBelthle — DISMISSED** [review]
The back-end side seems legit

File: webapp/src/components/App/Studies/StudyTree/index.tsx (hunk 18 above)
Comment (smailio) [reply webapp/src/components/App/Studies/StudyTree/index.tsx]: I want to write the unit test for insertWorkspacesIfNotExist it's easier if it's not coupled with the fetch

Comment (hdinia) [reply webapp/src/components/App/Studies/StudyTree/index.tsx]: ok 👍🏼

File: webapp/public/locales/en/main.json [hunk 23]
```diff
@@ -636,7 +636,9 @@
   "studies.copySuffix": "Copy",
   "studies.folder": "Folder",
   "studies.filters.strictfolder": "Show only direct folder children",
+  "studies.filters.showAllDescendants": "Show all children",
   "studies.scanFolder": "Scan folder",
+  "studies.requestDeepScan": "Scan the folder recursively (may be long) ?",
```
Comment (hdinia) [inline webapp/public/locales/en/main.json]:
```typescript
"studies.folder.recursiveScan": "Recursive scan"
"studies.folder.recursiveScan": "Scan récursif"
```
if really needed the time warning would be better in a Tooltip

File: webapp/public/locales/en/main.json [hunk 24]
```diff
@@ -636,7 +636,9 @@
   "studies.copySuffix": "Copy",
   "studies.folder": "Folder",
   "studies.filters.strictfolder": "Show only direct folder children",
+  "studies.filters.showAllDescendants": "Show all children",
```
Comment (hdinia) [inline webapp/public/locales/en/main.json:645]:
```typescript
"studies.folder.showChildren": "Show subfolders"
"studies.folder.showChildren": "Voir les sous-dossiers"
```

File: webapp/src/components/App/Studies/StudiesList/index.tsx [hunk 25]
```diff
@@ -91,6 +94,7 @@ function StudiesList(props: StudiesListProps) {
   const [selectedStudies, setSelectedStudies] = useState<string[]>([]);
   const [selectionMode, setSelectionMode] = useState(false);
   const [confirmFolderScan, setConfirmFolderScan] = useState<boolean>(false);
+  const [requestDeepScan, setRequestDeepScan] = useState<boolean>(false);
```
Comment (hdinia) [inline webapp/src/components/App/Studies/StudiesList/index.tsx]: Rename state variable for clarity:
- `requestDeepScan` -> `isRecursiveScan` (prefix "is" indicating a boolean flag)

Matches the API parameter name in `scanFolder()`

Also when you add a default state you dont have to specify the type, TS will infer it
`const [isRecursiveScan, setIsRecursiveScan] = useState(false);`

File: webapp/src/components/App/Studies/StudiesList/index.tsx [hunk 26]
```diff
@@ -252,13 +261,21 @@ function StudiesList(props: StudiesListProps) {
...
-          </Tooltip>
+
+          {strictFolderFilter ? (
+            <Tooltip title={t("studies.filters.strictfolder") as string}>
+              <IconButton onClick={toggleStrictFolder}>
+                <FolderIcon color="secondary" />
+              </IconButton>
+            </Tooltip>
+          ) : (
+            <Tooltip title={t("studies.filters.showAllDescendants") as string}>
```
Comment (hdinia) [inline webapp/src/components/App/Studies/StudiesList/index.tsx]: `as string` not necessary

File: webapp/src/components/App/Studies/StudiesList/index.tsx [hunk 27]
```diff
@@ -159,13 +163,18 @@ function StudiesList(props: StudiesListProps) {
...
-      await scanFolder(folder);
+      await scanFolder(folder, requestDeepScan);
       setConfirmFolderScan(false);
+      setRequestDeepScan(false);
     } catch (e) {
       enqueueErrorSnackbar(t("studies.error.scanFolder"), e as AxiosError);
     }
   };
 
+  const handleDeepScanCheckboxChange = () => {
```
Comment (hdinia) [inline webapp/src/components/App/Studies/StudiesList/index.tsx]: `handleDeepScanCheckboxChange` -> `handleRecursiveScan`

File: webapp/src/components/App/Studies/StudyTree/index.tsx (hunk 14 above)
Comment (hdinia) [reply webapp/src/components/App/Studies/StudyTree/index.tsx]: i18n key still missing, in addition please consider refactoring this event handler to use async/await syntax

**Review by hdinia — CHANGES_REQUESTED** [review]
Some minor feedback, otherwise it looks good to me 👍🏼

File: webapp/public/locales/fr/main.json [hunk 28]
```diff
@@ -667,7 +669,9 @@
   "studies.exportOutput": "Exporter une sortie",
   "studies.exportOutputFilter": "Exporter une sortie filtrée",
   "studies.selectOutput": "Selectionnez une sortie",
-  "studies.variant": "Variante",
```
Comment (skamril) [inline webapp/public/locales/fr/main.json:676]: Remove also the key in the english file

File: webapp/src/components/App/Studies/StudiesList/index.tsx [hunk 29]
```diff
@@ -269,12 +286,20 @@ function StudiesList(props: StudiesListProps) {
...
               onConfirm={handleFolderScan}
               alert="warning"
               open
             >
               {`${t("studies.scanFolder")} ${folder}?`}
+              <FormControlLabel
+                control={<Checkbox checked={isRecursiveScan} />}
+                label={t("studies.requestDeepScan")}
+                onChange={handleRecursiveScan}
+              />{" "}
```
Comment (skamril) [inline webapp/src/components/App/Studies/StudiesList/index.tsx]: Remove  `{" "}`

File: webapp/src/components/App/Studies/StudiesList/index.tsx [hunk 30]
```diff
@@ -269,12 +286,20 @@ function StudiesList(props: StudiesListProps) {
...
+              onCancel={() => {
+                setConfirmFolderScan(false);
+                setIsRecursiveScan(false);
+              }}
               onConfirm={handleFolderScan}
               alert="warning"
               open
             >
               {`${t("studies.scanFolder")} ${folder}?`}
+              <FormControlLabel
```
Comment (skamril) [inline webapp/src/components/App/Studies/StudiesList/index.tsx]: Use our component `CheckBoxFE`

File: webapp/src/components/App/Studies/StudyTree/index.tsx [hunk 31]
```diff
@@ -0,0 +1,146 @@
...
+          await fetchAndInsertWorkspaces(initialStudiesTree);
+        const childrenPaths = treeWithWorkspaces.children.map(
+          (child) => `root${child.path}`,
+        );
+        const [updatedTree, failedPaths] = await fetchAndInsertSubfolders(
+          childrenPaths,
+          treeWithWorkspaces,
+        );
+        setStudiesTree(updatedTree);
+        failedPaths.forEach((path) => {
```
Comment (skamril) [inline webapp/src/components/App/Studies/StudyTree/index.tsx]: This can flood the user with popups. Include them in one snackbar instead.

File: webapp/src/components/App/Studies/StudyTree/index.tsx [hunk 32]
```diff
@@ -0,0 +1,146 @@
...
+              interpolation: { escapeValue: false },
+            }),
+            t("studies.tree.error.detailsInConsole"),
+          );
+        });
+      } catch (error) {
+        setStudiesTree(initialStudiesTree);
+        enqueueErrorSnackbar(
+          t("studies.tree.error.failToFetchWorkspace"),
+          error as AxiosError,
```
Comment (skamril) [inline webapp/src/components/App/Studies/StudyTree/index.tsx]: Use `toError(err)`

File: webapp/src/components/App/Studies/StudyTree/fixtures.ts [hunk 33]
```diff
@@ -0,0 +1,157 @@
+/**
```
Comment (skamril) [inline webapp/src/components/App/Studies/StudyTree/fixtures.ts]: It's not better to have a test folder inside component @hdinia?

File: webapp/src/components/App/Studies/StudyTree/index.tsx [hunk 34]
```diff
@@ -0,0 +1,146 @@
...
+    for (const path of failedPath) {
+      enqueueErrorSnackbar(
+        t("studies.tree.error.failToFetchFolder", {
+          path,
+          interpolation: { escapeValue: false },
+        }),
+        t("studies.tree.error.detailsInConsole"),
+      );
+    }
+  };
```
Comment (skamril) [inline webapp/src/components/App/Studies/StudyTree/index.tsx:113]: Add a line break

File: webapp/src/components/App/Studies/StudyTree/index.tsx [hunk 35]
```diff
@@ -0,0 +1,146 @@
...
+        const nextTree = await fetchAndInsertWorkspaces(studiesTree);
+        setStudiesTree(nextTree);
+      } catch (error) {
+        enqueueErrorSnackbar(
+          "studies.tree.error.failToFetchWorkspace",
+          error as AxiosError,
+        );
+      }
+    }
+    const chidrenPaths = studyTreeNode.children.map(
```
Comment (skamril) [inline webapp/src/components/App/Studies/StudyTree/index.tsx]: This code and the code from the effect can be move in a common function

**Review by hdinia — DISMISSED** [review]

File: webapp/src/components/App/Studies/StudyTree/utils.ts [hunk 36]
```diff
@@ -0,0 +1,196 @@
...
+  );
+}
+
+/**
+ * Call the explorer api to fetch the subfolders under the given path.
+ *
+ * @param path path of the subfolder to fetch, should sart with root, e.g. root/workspace/folder1
+ * @returns list of subfolders under the given path
+ */
+async function fetchSubfolders(path: string): Promise<NonStudyFolderDTO[]> {
```
Comment (skamril) [inline webapp/src/components/App/Studies/StudyTree/utils.ts:100]: `async / await` and return type are not necessary

**Review by skamril — CHANGES_REQUESTED** [review]

File: webapp/src/components/App/Studies/StudyTree/fixtures.ts (hunk 33 above)
Comment (hdinia) [reply webapp/src/components/App/Studies/StudyTree/fixtures.ts]: indeed to not mix concerns if their is multiple files related to tests in a folder we should put them inside their `__tests__` folder

**Review by skamril — DISMISSED** [review]

File: webapp/public/locales/fr/main.json (hunk 28 above)
Comment (smailio) [reply webapp/public/locales/fr/main.json:676]: I think it was a duplicate in tis file or something

File: webapp/src/components/App/Studies/StudyTree/index.tsx (hunk 34 above)
Comment (smailio) [reply webapp/src/components/App/Studies/StudyTree/index.tsx:113]: done

**Review by hdinia — APPROVED** [review]
